docs/design-notes/qualifying-replicas-for-reads.md

```diff
@@ -0,0 +1,69 @@
+# Qualifying replicas for read operations
+
+## Intro
+
+When a read request arrives at the coordinator node, it is later
+forwarded to one or more replicas in order to read the actual data
+and return it back to the client. Each request has an associated
+timeout, which indicates how long the client is willing to wait
+for the query results.
+This timeout information can be leveraged to lift some pressure off
+a replica which is overloaded or dead.
+
+## Gathering last seen information
+
+The coordinator should be able to predict whether a replica is likely
+to respond to a request in time or not. This prediction should be as
+precise as possible, but at the same time have minimal overhead, because
+performing read requests is on the hot path. A simple estimation
+of how responsive a replica is likely to be is implemented as follows:
+ * each time a read request is sent to a replica, a timestamp is recorded
+ * each time a response is received from a replica, a timestamp is recorded
+ * the difference between these two timestamps is a cheap, rough estimate
+   of how long it may take this replica to respond
+Note that this estimation is not precise - it does not determine whether
+a `sent` timestamp and a `responded` timestamp belong to the same request,
+so these values may vary. However, assuming that the cluster is under
+continuous, uniform load, these values would get updated very often, and would
+hopefully approach the approximate round-trip time for a single request.
+Finally, when a replica was not contacted in a long while (e.g. 5s),
+its previous last seen information is considered stale and dropped to avoid
+reporting incorrect long periods of time without response.
+
+## Omitting replicas not likely to respond in time
+
+Based on the last seen information, the coordinator is now allowed
+to make an educated guess whether a replica is going to respond in time.
+A simple, probabilistic approach is implemented as follows.
+At first, we compute how much time left (`TL`) the request still has until
```

Comment:
This number is also dangerous. Usually it's similar to the timeout. But if the coordinator is overloaded, it gets small (or even zero or negative). If it appears in the denominator of a division, it gets amplified.

Reply:
Right. In the denominator, it would perhaps make sense to use the timeout duration, which is less prone to getting really small (unless explicitly specified by the user). That one would need to be passed everywhere explicitly, but it's lightweight.

Reply:
It also feels fishy. It makes a workload that has low timeouts be very sensitive to extra delays. These delays could ripple (a problem with dynamic snitch): a node gets a little slow, other cooridnators back off and send requests to another node; that node becomes slow (and the original node is now fast) and traffic flip-flops between the two.

I think the failure detector part in this patch works and was tested, and the overload backoff wasn't.